File: ruby/lib/resp_bench/engine/command_selector.rb
# frozen_string_literal: true

module RespBench
  module Engine
    # Selects commands based on their configured weights.
    # Uses cumulative weights for efficient O(log n) selection.
    class CommandSelector
      # @param commands [Array<RespBench::Command::Command>] list of commands with weights
      def initialize(commands)
        @commands = commands
        @cumulative_weights = build_cumulative_weights(commands)
        @random = Random.new # Ruby's built-in random is fine for selection
      end

      # Select a command based on configured weights
      # @return [RespBench::Command::Command] selected command
      def select
        r = @random.rand
        @cumulative_weights.each_with_index do |threshold, index|
          return @commands[index] if r <= threshold
        end
        @commands.last
      end

      private

      def build_cumulative_weights(commands)
        # Normalize weights to sum to 1.0
        total_weight = commands.sum(&:weight)
        total_weight = 1.0 if total_weight.zero?

        cumulative = []
        sum = 0.0
        commands.each do |cmd|
          sum += cmd.weight / total_weight
          cumulative << sum
        end
        cumulative
      end
    end
  end
end

File: python/src/resp_bench/engine/command_selector.py
"""Weighted command selection.

Uses normalized cumulative weights. Command selection intentionally uses
Python's built-in RNG (not the Java LCG): only key generation must be
cross-engine deterministic, command selection need not be.
"""

from __future__ import annotations

import random
from typing import List

from ..command.command import Command


class CommandSelector:
    def __init__(self, commands: List[Command]) -> None:
        self._commands = commands
        self._cumulative_weights = self._build_cumulative_weights(commands)
        self._random = random.Random()

    def select(self) -> Command:
        r = self._random.random()
        for index, threshold in enumerate(self._cumulative_weights):
            if r <= threshold:
                return self._commands[index]
        return self._commands[-1]

    @staticmethod
    def _build_cumulative_weights(commands: List[Command]) -> List[float]:
        total_weight = sum(c.weight for c in commands)
        if total_weight == 0:
            total_weight = 1.0

        cumulative: List[float] = []
        running = 0.0
        for command in commands:
            running += command.weight / total_weight
            cumulative.append(running)
        return cumulative
